jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java

```diff
@@ -0,0 +1,124 @@
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.jsr352.jberet.impl;
+
+import java.lang.annotation.Annotation;
+
+import javax.enterprise.context.ContextNotActiveException;
+import javax.enterprise.context.spi.Context;
+import javax.enterprise.context.spi.Contextual;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.AfterBeanDiscovery;
+import javax.enterprise.inject.spi.AfterTypeDiscovery;
+import javax.enterprise.inject.spi.AnnotatedType;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.Extension;
+
+import org.hibernate.search.jsr352.context.jpa.EntityManagerFactoryRegistry;
+import org.hibernate.search.jsr352.inject.scope.HibernateSearchJobScoped;
+import org.hibernate.search.jsr352.inject.scope.HibernateSearchPartitionScoped;
+import org.hibernate.search.jsr352.massindexing.impl.JobContextSetupListener;
+import org.hibernate.search.jsr352.massindexing.impl.steps.lucene.EntityReader;
+import org.jberet.cdi.JobScoped;
+import org.jberet.cdi.PartitionScoped;
+
+
+/**
+ * A CDI extension to explicitly add hibernate-search-jsr352-core types to the CDI context.
+ * <p>
+ * These types must be registered into the CDI context because they expect one of their
+ * attributes to be injected, thus they need to be instantiated by CDI.
+ * <p>
+ * Ideally we should use component scanning to that effect, but:
+ * <ul>
+ * <li>In CDI, component scanning is configured per-module in META-INF/beans.xml,
+ * with no way to add a package from another module.
+ * <li>In Wildfly, component scanning in dependencies is always done with the "all" discovery type,
+ * meaning that even components without CDI annotations will be discovered.
+ * Besides exposing private types unnecessarily,
+ * this is annoying because it puts default {@link EntityManagerFactoryRegistry} implementations into the CDI context,
+ * leading to conflicts when injecting such types
+ * (since this module, hibernate-search-jsr352-jberet, also provides an implementation).
+ * <br>See <a href="https://issues.jboss.org/browse/WFLY-8656">WFLY-8656</a>.
+ * <li>Even if we could solve the above, JBeret requires specific scope annotations to be set on the
+ * beans so that they are correctly scoped. Since the core isn't dependent on JBeret, we must specify these
+ * scopes manually when registering the beans.
+ * </ul>
+ * Thus we use explicit type and scope registration as a workaround.
+ *
+ * @author Yoann Rodiere
+ */
+public class HibernateSearchJsr352Extension implements Extension {
+
+	public void afterTypeDiscovery(@Observes AfterTypeDiscovery event, BeanManager beanManager) {
+		registerType( event, beanManager, JobContextSetupListener.class );
+		registerType( event, beanManager, EntityReader.class );
```

Comment:
> Such operation is necessary because we want to inject value into attributes of these classes. Therefore, they need to be instantiated before injection. For example, we want to inject the EntityManagerFactoryRegistry into the job setup listener, so the listener must be instantiated first.

Right. CDI only injects beans that it created itself. When JBeret doesn't find a bean in the CDI context, it instantiates the bean itself, and thus the bean is not injected. So we really need the bean to be known from CDI and instantiated by CDI.

> Normally, during CDI component scanning, the annotated bean can be discovered, register (instantiate) correctly. However, it isn't the case in WildFly (see WFLY-8656). So we have to create this extension class to handle the stuff manually.

Right. To be precise, it *is* the case in WildFly (the beans are discovered), but WildFly is much too aggressive: it will discover not only the annotated beans, but *any* bean, even those we don't want to be in the CDI context. So we work it around by disabling automatic discovery and by registering the types manually.

> Am I right?

Yes.